src/api/VoteMonitor.Api.Form/CommandHandlers/DeleteFormCommandHandler.cs

```diff
@@ -0,0 +1,55 @@
+﻿using MediatR;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using VoteMonitor.Api.Form.Commands;
+using VoteMonitor.Entities;
+
+namespace VoteMonitor.Api.Form.CommandHandlers
+{
+    public class DeleteFormCommandHandler : IRequestHandler<DeleteFormCommand, bool>
+    {
+        private readonly VoteMonitorContext _context;
+
+        public DeleteFormCommandHandler(VoteMonitorContext context)
+        {
+            this._context = context;
+        }
+
+        public async Task<bool> Handle(DeleteFormCommand request, CancellationToken cancellationToken)
+        {
+            var form = await _context.Forms.FirstOrDefaultAsync(f => f.Id == request.FormId);
+            if (form == null)
+            {
+                return false;
+            }
+
+            var sections = _context.FormSections.Where(s => s.IdForm == form.Id);
+            var sectionsIds = sections.Select(s => s.Id);
+            var questions = _context.Questions.Where(q => sectionsIds.Contains(q.IdSection));
+            var questionsIds = questions.Select(q => q.Id);
+            var optionsToQuestions = _context.OptionsToQuestions.Where(o => questionsIds.Contains(o.IdQuestion));
+            var optionsIds = optionsToQuestions.Select(o => o.IdOption);
+
+            // check if there are already saved answers
+            var answers = _context.Answers.Where(a => optionsIds.Contains(a.IdOptionToQuestion));
+            if (answers != null && answers.Any())
+            {
+                return false;
+            }
+
+            var options = _context.Options.Where(o => optionsIds.Contains(o.Id));
+
+            _context.OptionsToQuestions.RemoveRange(optionsToQuestions);
```

Comment:
`RemoveRange` materializes the query so in order to avoid blocking here suggest fetching data asynchronously, thus something like `RemoveRange(await optionsToQuestions.ToListAsync())`. Same for the other calls

Reply:
In that case, the collection will be enumerated twice(once by ToListAsync and once by RemoveRange. I propose to keep it like this(this was the way it was before too) and add a task to not enumerate collection at all by creating some extension methods to delete a range of entities using SqlCommands(get the SQL command to generate by the select and remove it correctly). I can take that task after this one. What do you say?

Reply:
optionsToQuestions is an IQueryable so this is why RemoveRange know it has to materialize it.
toList create a list and there is no way to reach the iqueryable so this is why the query is sent to the server only once

Reply:
btw, we could check if we actually see the queries (enable info logging for Microsoft on your appsettings.dev)

Reply:
I know that _optionsToQuestions_ is IQueryable and I agree the query will be executed only once with your approach(when ToListAsync is evaluated). But as far as I know, that list will be enumerated one more time by RemoveRange method.
Db => ToList (One Query + One Enumeration)
List => RemoveRange(One Enumeration)
I had this problem before on some projects with a large amount of data and the best option is to not enumerate or query the data at all(ef core doesn't support that by default, but can easily archived by visiting the expression in order to compose the query, alter the query to delete instead of select, and execute the query directly.

Reply:
Fine with me then, I didn't know about this issue.